### PR TITLE
[BugFix] Optimize single table cost calculation

### PR DESCRIPTION
**Pull request by ABingHuang**
Optimize single table cost calculation, choose the final plan based on the mv's output rows, column compute size, instead of uncorrected estimated statistics.

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: test/sql/test_materialized_view/T/test_materialized_view_rewrite [hunk 1]
```diff
@@ -264,4 +264,36 @@ explain select depts.deptno
...
+distributed by hash(user_id)
+as
+select user_id, user_name, count(1) as cnt
+from user_tags
+group by user_id, user_name;
+
+refresh materialized view agg_count_mv2 with sync mode;
+analyze table agg_count_mv2 with sync mode;
+
+explain select user_id, count(1) as cnt
```
Comment (kangkaisen) [inline test/sql/test_materialized_view/T/test_materialized_view_rewrite:297]: we should check the explain result?

File: test/sql/test_materialized_view/T/test_materialized_view_rewrite [hunk 2]
```diff
@@ -264,4 +264,36 @@ explain select depts.deptno
...
+
+create materialized view agg_count_mv1
+distributed by hash(user_id)
+as
+select user_id, count(1) as cnt
+from user_tags
+group by user_id;
+
+refresh materialized view agg_count_mv1 with sync mode;
+analyze table agg_count_mv1 with sync mode;
```
Comment (kangkaisen) [inline test/sql/test_materialized_view/T/test_materialized_view_rewrite:285]: For the bug case, I think we don't need to analyze.

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 28 / 30 (93.33%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/rule/SingleTableRewriteBaseRule.java|28|30|93.33%|[126, 127]|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/rule/SingleTableRewriteBaseRule.java [hunk 3]
```diff
@@ -45,9 +49,48 @@ public List<OptExpression> transform(OptExpression queryExpression, OptimizerCon
...
+        }
+    }
+
+    @VisibleForTesting
+    public static class CandidateContextComparator implements Comparator<CandidateContext> {
+        @Override
+        public int compare(CandidateContext context1, CandidateContext context2) {
+            int ret = Double.compare(context1.getMvStatistics().getOutputRowCount(),
+                    context2.getMvStatistics().getOutputRowCount());
+            if (ret != 0) {
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/rule/SingleTableRewriteBaseRule.java:89]: This isn't right. for the query group by a, MV1: group by a,b, MV2:group by a, we should always choose MV2.

Comment (satanson) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/rule/SingleTableRewriteBaseRule.java:128]: there is an anti-example:

  1. count(1) group by a,b  is slightly larger than count(1) group by a;
  2. column a's distribution skews drastically, but column group (a,b) 's distribution is smooth.

**Review by kangkaisen — APPROVED** [review]

**Review by LiShuMing — APPROVED** [review]

**Review by satanson — APPROVED** [review]

**Pull request merged**